server/tests/cluster.rs

```diff
@@ -76,3 +82,170 @@ async fn replicated_loglet() -> googletest::Result<()> {
+        .build()
+        .expect("reqwest client should build");
+
+    let discovery_response = client
+        .post(format!("http://{}/deployments", admin_address))
+        .header(CONTENT_TYPE, "application/json")
+        .body(
+            serde_json::json!({"uri": format!("http://{}", service_endpoint_address)}).to_string(),
+        )
+        .send()
+        .await?;
+
+    assert!(
+        discovery_response.status().is_success(),
+        "discovery should be successful"
+    );
+
+    info!("Successfully registered service endpoint at cluster");
+
+    let start_chaos = Instant::now();
+
+    let chaos_handle = TaskCenter::spawn_unmanaged(TaskKind::TestRunner, "chaos", async move {
+        let mut shutdown = std::pin::pin!(cancellation_watcher());
+
+        loop {
+            let node = cluster
+                .nodes
+                .choose_mut(&mut rand::rng())
+                .expect("at least one node being present");
+
+            tokio::select! {
+                _ = &mut shutdown => {
+                    break;
+                },
+                result = node.restart() => {
+                    result?;
+                    // wait until the cluster is healthy again
+                    tokio::select! {
+                        _ = &mut shutdown => {
+                            break;
+                        }
+                        result = cluster.wait_healthy(Duration::from_secs(10)) => {
+                            result?;
+                        }
+                    }
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                }
+            }
+        }
+
+        Ok::<_, anyhow::Error>(cluster)
+    })?;
+
+    info!("Starting the cluster chaos test");
+
+    let mut successes = 0;
+    let mut failures = 0;
+    let mut rng = rand::rng();
+    while start_chaos.elapsed() < chaos_duration {
+        let ingress = ingress_addresses
+            .choose(&mut rng)
+            .expect("at least one address to be present");
+
+        match client
+            .post(ingress)
+            .header(CONTENT_TYPE, "application/json")
+            .header("idempotency-key", successes.to_string())
+            .body("1")
+            .send()
+            .await
+        {
+            Ok(response) => {
+                if response.status().is_success() {
+                    successes += 1;
+                    let counter_value: i32 =
+                        serde_json::from_slice(response.bytes().await?.as_ref())?;
+                    assert_eq!(successes, counter_value);
+                } else {
+                    failures += 1;
+                }
+            }
+            Err(_) => {
+                failures += 1;
+                // request failed, let's retry
+            }
+        }
+    }
+
+    // make sure that we have written at least some values
+    assert!(
+        successes > 10,
+        "successful writes: {successes} failed writes: {failures}",
+    );
```

Comment:
Maybe out of scope for what you intended here, but a nice enhancement for the future might be to pause killing nodes, and assert that some successes happened during the final restart cycle. Maybe even after every single restart, wait for both cluster healthy and some successful responses to come back. To be clear, not a blocker, just thinking how we could make this test catch more problems for the same "cost".

Reply:
I think I can add it if I expect a certain time interval within which the cluster needs to recover. Let's see whether this is too strict for the test to be stable.